// stunexus/src/context/AuthContext.jsx
import React, { createContext, useState, useEffect } from 'react';
import { jwtDecode } from 'jwt-decode';
import axiosClient from '../utils/axiosClient';

export const AuthContext = createContext();

export const AuthProvider = ({ children }) => {
  const [user, setUser] = useState(null);
  const [loading, setLoading] = useState(true);

  // Hàm giải mã Token JWT -> lấy ra thông tin user
  const decodeAndSetUser = (token, role) => {
    try {
      const decoded = jwtDecode(token);
      // Claims trong JWT của BE:
      // sub = TaiKhoan, http://...nameidentifier = MaGV/MaSV/MaQTV, http://...name = HoTen, http://...role = Role
      const userData = {
        TaiKhoan: decoded.sub,
        MaId: decoded['http://schemas.xmlsoap.org/ws/2005/05/identity/claims/nameidentifier'],
        HoTen: decoded['http://schemas.xmlsoap.org/ws/2005/05/identity/claims/name'],
        role: role || decoded['http://schemas.microsoft.com/ws/2008/06/identity/claims/role'],
      };
      
      // Map role về đúng chuẩn FE đang dùng
      if (userData.role === 'Admin') userData.role = 'admin';
      else if (userData.role === 'Lecturer') userData.role = 'lecturer';
      else if (userData.role === 'Student') userData.role = 'student';

      // Gán mã ID cụ thể theo role
      if (userData.role === 'admin') userData.MaQTV = userData.MaId;
      else if (userData.role === 'lecturer') userData.MaGV = userData.MaId;
      else if (userData.role === 'student') userData.MaSV = userData.MaId;

      setUser(userData);
      localStorage.setItem('stu_user', JSON.stringify(userData));
      return userData;
    } catch (err) {
      console.error('Giải mã Token thất bại:', err);
      return null;
    }
  };

  // Khôi phục session từ localStorage khi F5
  useEffect(() => {
    const token = localStorage.getItem('token');
    if (token) {
      try {
        const decoded = jwtDecode(token);
        // Kiểm tra Token hết hạn chưa
        if (decoded.exp * 1000 > Date.now()) {
          const stored = localStorage.getItem('stu_user');
          if (stored) {
            setUser(JSON.parse(stored));
          } else {
            decodeAndSetUser(token);
          }
        } else {
          // Token hết hạn -> dọn sạch
          localStorage.removeItem('token');
          localStorage.removeItem('stu_user');
        }
      } catch (err) {
        localStorage.removeItem('token');
        localStorage.removeItem('stu_user');
      }
    }
    setLoading(false);
  }, []);

  // Hàm Login: Gọi API thật từ Backend C#
  const login = async (username, password) => {
    try {
      const response = await axiosClient.post('/auth/login', {
        taiKhoan: username,
        matKhau: password
      });

      if (response.data && response.data.success) {
        const token = response.data.data.token;
        const role = response.data.data.role;

        // Cất Token vào két sắt localStorage
        localStorage.setItem('token', token);

        // Giải mã Token và set User session
        const userData = decodeAndSetUser(token, role);
        
        if (userData) {
          return { success: true, role: userData.role };
        }
      }
      return { success: false, message: response.data?.message || 'Đăng nhập thất bại' };
    } catch (err) {
      console.error('Login error:', err);
      const msg = err.response?.data?.message || 'Sai tài khoản hoặc mật khẩu!';
      return { success: false, message: msg };
    }
  };

  const logout = () => {
    setUser(null);
    localStorage.removeItem('stu_user');
    localStorage.removeItem('token');
  };

  const updateUserSession = (newUserData) => {
    const updated = { ...user, ...newUserData };
    setUser(updated);
    localStorage.setItem('stu_user', JSON.stringify(updated));
  };

  return (
    <AuthContext.Provider value={{ user, login, logout, updateUserSession, loading }}>
      {children}
    </AuthContext.Provider>
  );
};
